data-pipeline/src/data_pipeline/datasets/clinvar.py

```diff
@@ -47,6 +47,77 @@ def escape_xpath_value(value):
+    if trait_mapping_list_element is None:
+        return None
+
+    xref_elements = trait_element.findall("XRef")
+    for xref_element in xref_elements:
+        selector = f"./TraitMapping[@ClinicalAssertionID='{submission_element.attrib['ID']}'][@TraitType='{trait_element.attrib['Type']}'][@MappingType='XRef'][@MappingValue='{escape_xpath_value(xref_element.attrib['ID'])}']"  # noqa
+        mapping_element = trait_mapping_list_element.find(selector)
+        if mapping_element is not None:
+            return mapping_element
+    return None
+
+
+def find_mapping_elements_by_preferred_name(trait_element, submission_element, trait_mapping_list_element):
+    preferred_name_element = trait_element.find("./Name/ElementValue[@Type='Preferred']")
+
+    if preferred_name_element is not None and trait_mapping_list_element is not None:
+        selector = f"./TraitMapping[@ClinicalAssertionID='{submission_element.attrib['ID']}'][@TraitType='{trait_element.attrib['Type']}'][@MappingType='Name'][@MappingValue='{escape_xpath_value(preferred_name_element.text)}']"  # noqa
+        mapping_element = trait_mapping_list_element.find(selector)
+        return mapping_element, preferred_name_element
+
+    return None, preferred_name_element
+
+
+def find_mapping_elements_by_name(trait_element, submission_element, trait_mapping_list_element):
+    name_elements = trait_element.findall("./Name/ElementValue")
+    preferred_name_element = None
+
+    for name_element in name_elements:
+        if preferred_name_element is None:
+            preferred_name_element = name_element
+
+        if trait_mapping_list_element is not None:
+            selector = f"./TraitMapping[@ClinicalAssertionID='{submission_element.attrib['ID']}'][@TraitType='{trait_element.attrib['Type']}'][@MappingType='Name'][@MappingValue='{escape_xpath_value(name_element.text)}']"  # noqa
+            mapping_element = trait_mapping_list_element.find(selector)
+            if mapping_element:
+                return mapping_element, preferred_name_element
+
+    return None, preferred_name_element
+
+
+def _determine_mapping_and_preferred_name_element(trait_element, submission_element, trait_mapping_list_element):
+    preferred_name_element = None
+    mapping_element = find_mapping_elements_by_xref(trait_element, submission_element, trait_mapping_list_element)
+
+    if mapping_element is None:
+        mapping_element, preferred_name_element = find_mapping_elements_by_preferred_name(
+            trait_element, submission_element, trait_mapping_list_element
+        )
+
+    if mapping_element is None:
+        mapping_element, preferred_name_element = find_mapping_elements_by_name(
+            trait_element, submission_element, trait_mapping_list_element
+        )
+
+    return (mapping_element, preferred_name_element)
+
+
+def _associate_condition_with_medgen_id(submission_element, trait_mapping_list_element, trait_element):
```

Comment:
Nice one, you see how both the overall process and the individual parts are both easier to understand now.